File: UngDungDiemDanhNhanVien/DTOs/DangKyRequest.cs
using System.ComponentModel.DataAnnotations;

namespace UngDungDiemDanhNhanVien.DTOs
{
    public class DangKyRequest
    {
        [Required(ErrorMessage = "Vai trò là bắt buộc")]
        [StringLength(50, ErrorMessage = "Vai trò không được quá 50 ký tự")]
        public string VaiTro { get; set; } = string.Empty; // Admin, QuanLy, NhanVien

        // Thông tin chung
        [Required(ErrorMessage = "Email là bắt buộc")]
        [EmailAddress(ErrorMessage = "Email không hợp lệ")]
        [StringLength(100, ErrorMessage = "Email không được quá 100 ký tự")]
        public string Email { get; set; } = string.Empty;

        [StringLength(15, ErrorMessage = "Số điện thoại không được quá 15 ký tự")]
        [Phone(ErrorMessage = "Số điện thoại không hợp lệ")]
        public string? SoDienThoai { get; set; }

        // Thông tin Admin/QuanLy
        [StringLength(50, MinimumLength = 3, ErrorMessage = "Tên đăng nhập phải từ 3-50 ký tự")]
        public string? TenDangNhap { get; set; }

        [StringLength(100, MinimumLength = 6, ErrorMessage = "Mật khẩu phải từ 6-100 ký tự")]
        public string? MatKhau { get; set; }

        [Compare("MatKhau", ErrorMessage = "Mật khẩu và xác nhận mật khẩu không khớp")]
        public string? XacNhanMatKhau { get; set; }

        // Thông tin Nhân viên
        [StringLength(20, MinimumLength = 3, ErrorMessage = "Mã nhân viên phải từ 3-20 ký tự")]
        public string? MaNhanVien { get; set; }

        [StringLength(100, MinimumLength = 2, ErrorMessage = "Họ tên phải từ 2-100 ký tự")]
        public string? HoTen { get; set; }

        // Mật khẩu cho nhân viên
        [StringLength(100, MinimumLength = 6, ErrorMessage = "Mật khẩu nhân viên phải từ 6-100 ký tự")]
        public string? MatKhauNhanVien { get; set; }

        [StringLength(100, ErrorMessage = "Phòng ban không được quá 100 ký tự")]
        public string? PhongBan { get; set; }

        [StringLength(100, ErrorMessage = "Chức vụ không được quá 100 ký tự")]
        public string? ChucVu { get; set; }

        [Range(0, double.MaxValue, ErrorMessage = "Lương giờ phải lớn hơn 0")]
        public decimal? LuongGio { get; set; }

        [StringLength(20, ErrorMessage = "Trạng thái không được quá 20 ký tự")]
        public string TrangThai { get; set; } = "HoatDong";

        [StringLength(50, ErrorMessage = "Mã sinh trắc học không được quá 50 ký tự")]
        public string? MaSinhTracHoc { get; set; }

        [StringLength(50, ErrorMessage = "Mã khuôn mặt không được quá 50 ký tự")]
        public string? MaKhuonMat { get; set; }
    }
}

File: ung_dung_diem_danh/lib/models/dang_ky_request.dart
class DangKyRequest {
  final String tenDangNhap;
  final String matKhau;
  final String email;
  final String vaiTro;
  final String? hoTen;
  final String? soDienThoai;
  final String? xacNhanMatKhau;

  DangKyRequest({
    required this.tenDangNhap,
    required this.matKhau,
    required this.email,
    required this.vaiTro,
    this.hoTen,
    this.soDienThoai,
    String? xacNhanMatKhau,
  }) : xacNhanMatKhau = xacNhanMatKhau ?? matKhau; // Auto-fill với matKhau nếu null

  Map<String, dynamic> toJson() {
    return {
      'tenDangNhap': tenDangNhap,
      'matKhau': matKhau,
      'email': email,
      'vaiTro': vaiTro,
      'hoTen': hoTen,
      'soDienThoai': soDienThoai,
      'xacNhanMatKhau': xacNhanMatKhau,
    };
  }

  factory DangKyRequest.fromJson(Map<String, dynamic> json) {
    return DangKyRequest(
      tenDangNhap: json['tenDangNhap'] ?? '',
      matKhau: json['matKhau'] ?? '',
      email: json['email'] ?? '',
      vaiTro: json['vaiTro'] ?? '',
      hoTen: json['hoTen'],
      soDienThoai: json['soDienThoai'],
      xacNhanMatKhau: json['xacNhanMatKhau'],
    );
  }
}

class DangKyAdminRequest extends DangKyRequest {
  DangKyAdminRequest({
    required String tenDangNhap,
    required String matKhau,
    required String email,
    String? hoTen,
    String? soDienThoai,
  }) : super(
          tenDangNhap: tenDangNhap,
          matKhau: matKhau,
          email: email,
          vaiTro: 'Admin',
          hoTen: hoTen,
          soDienThoai: soDienThoai,
        );
}

class DangKyNhanVienRequest extends DangKyRequest {
  final String maNhanVien;
  final String? phongBan;
  final String? chucVu;
  final double luongGio;

  DangKyNhanVienRequest({
    required this.maNhanVien,
    required String tenDangNhap,
    required String matKhau,
    required String email,
    required this.luongGio,
    String? hoTen,
    String? soDienThoai,
    this.phongBan,
    this.chucVu,
  }) : super(
          tenDangNhap: tenDangNhap,
          matKhau: matKhau,
          email: email,
          vaiTro: 'NhanVien',
          hoTen: hoTen,
          soDienThoai: soDienThoai,
        );

  @override
  Map<String, dynamic> toJson() {
    final json = super.toJson();
    json.addAll({
      'maNhanVien': maNhanVien,
      'phongBan': phongBan,
      'chucVu': chucVu,
      'luongGio': luongGio,
    });
    return json;
  }

  factory DangKyNhanVienRequest.fromJson(Map<String, dynamic> json) {
    return DangKyNhanVienRequest(
      maNhanVien: json['maNhanVien'] ?? '',
      tenDangNhap: json['tenDangNhap'] ?? '',
      matKhau: json['matKhau'] ?? '',
      email: json['email'] ?? '',
      luongGio: (json['luongGio'] ?? 0.0).toDouble(),
      hoTen: json['hoTen'],
      soDienThoai: json['soDienThoai'],
      phongBan: json['phongBan'],
      chucVu: json['chucVu'],
    );
  }
}

